libfaasmpi/CMakeLists.txt

```diff
@@ -16,50 +16,56 @@ set(LIB_FILES
     faasmpi.cpp
 )
 
-# Only do the install if we're doing a WASM build
-# WARNING - this overwrites the existing mpi.h on the system if not used in a
-# wasm build
-if (CMAKE_SYSTEM_NAME STREQUAL "Wasm")
-    message(STATUS "Libfaasmpi WebAssembly build")
+add_library(faasmpi STATIC ${LIB_FILES})
+set_target_properties(faasmpi PROPERTIES PUBLIC_HEADER "${PUBLIC_HEADERS}")
 
-    add_library(faasmpi STATIC ${LIB_FILES})
-    set_target_properties(faasmpi PROPERTIES PUBLIC_HEADER "${PUBLIC_HEADERS}")
+# Install Faabric MPI library
+set(FAABRIC_SOURCE_DIR ${CMAKE_CURRENT_LIST_DIR}/../third-party/faabric/src/mpi)
+set(FAABRIC_BINARY_DIR ${FAABRIC_SOURCE_DIR}/build)
 
-    # Add Faabric dependency
-    set(FAABRIC_SOURCE_DIR ${CMAKE_BINARY_DIR}/../../third-party/faabric/src/mpi)
-    set(FAABRIC_BINARY_DIR ${FAABRIC_SOURCE_DIR}/build)
-    add_subdirectory(${FAABRIC_SOURCE_DIR} ${FAABRIC_BINARY_DIR})
-    add_dependencies(faasmpi faabricmpi)
-    install(FILES ${FAABRIC_BINARY_DIR}/libfaabricmpi.a
-            DESTINATION ${CMAKE_SYSROOT}/lib/wasm32-wasi
-    )
+add_subdirectory(${FAABRIC_SOURCE_DIR} ${FAABRIC_BINARY_DIR})
+add_dependencies(faasmpi faabricmpi)
+install(FILES ${FAABRIC_BINARY_DIR}/libfaabricmpi.a
+        DESTINATION ${CMAKE_SYSROOT}/lib/wasm32-wasi
+)
+
+if (CMAKE_SYSTEM_NAME STREQUAL "Wasm")
+    message(STATUS "Libfaasmpi WebAssembly build")
 
-    # Install in sysroot
     install(TARGETS faasmpi
             ARCHIVE DESTINATION ${CMAKE_SYSROOT}/lib/wasm32-wasi
             LIBRARY DESTINATION ${CMAKE_SYSROOT}/lib/wasm32-wasi
             PUBLIC_HEADER DESTINATION ${CMAKE_SYSROOT}/include/faasmpi
     )
 
-    # Add imports
+    # Add imports file
     install(FILES faasmpi.imports
             DESTINATION ${CMAKE_SYSROOT}/lib/wasm32-wasi
             RENAME libfaasmpi.imports
     )
 
-    # Symlink header
-    install(CODE "execute_process( \
-        COMMAND ${CMAKE_COMMAND} -E create_symlink \
-        ${CMAKE_SYSROOT}/include/faabric/mpi/mpi.h \
-        ${CMAKE_SYSROOT}/include/mpi.h   \
-        )"
-    )
-
-    # Symlink library
+    # Symlink library as libmpi
     install(CODE "execute_process( \
         COMMAND ${CMAKE_COMMAND} -E create_symlink \
         ${CMAKE_SYSROOT}/lib/wasm32-wasi/libfaabricmpi.a \
         ${CMAKE_SYSROOT}/lib/wasm32-wasi/libmpi.a   \
         )"
     )
+else()
+    message(STATUS "Libfaasmpi native build")
+
+    install(TARGETS faasmpi
+            ARCHIVE DESTINATION lib
+            LIBRARY DESTINATION lib
+            PUBLIC_HEADER DESTINATION include/faasmpi
+            )
 endif()
+
+# Symlink header (common to wasm and native builds)
+# WARNING - this may overwrite the existing mpi.h on the system 
```

Comment:
Are you sure we want to still do this symlink if we are doing a native build? I see the point here, but could it ever be that in a native build two `MPI` implementations co-exist? (Namely `faabric`'s and a standard one?)

Or are we deliberately limiting the existing MPI implementation to use the subset of calls defined by `faabric`?

Reply:
Why can't this co-exist with another native MPI implementation? The symlinking is only performed when someone calls `ninja/make install`, and respects `CMAKE_INSTALL_PREFIX` to allow installing in different locations (i.e. I could install openmpi at `/usr/local/include` and this library at `/usr/local/faasm/faasmpi/include` if I wanted to.

The important thing for us is that we can natively compile Faasm functions that have `#include <mpi.h>` but do so against _our_ MPI implementation rather than something like OpenMPI.

Reply:
What has become annoying is that because the installed library doesn't live in `/usr/local/faasm`, it's difficult to copy in a multi-stage build, so I will probably change it to install at `/usr/local/faasm/native/faasmpi` for example.